src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -598,6 +663,24 @@ private static Expression BindParameterFromValue(ParameterInfo parameter, Expres
 
             var tryParseCall = tryParseMethodCall(parsedValue);
 
+            // The following code is generated if the parameter is required and
+            // the method should not be matched.
+            //
+            // if (tempSourceString == null)
+            // {
+            //      wasTryParseFailure = true;
+            //      Log.RequiredParameterNotProvided(httpContext, "Int32", "param1");    
+            // }
+            var checkRequiredParaseableParameterBlock = Expression.Block(
+                Expression.IfThen(TempSourceStringNullExpr,
+                    Expression.Block(
+                        Expression.Assign(WasTryParseFailureExpr, Expression.Constant(true)),
+                        Expression.Call(LogRequiredParameterNotProvidedMethod,
+                            HttpContextExpr, parameterTypeNameConstant, parameterNameConstant)
+                    )
+                )
+            );
```

Comment:
I'd prefer this was extracted into a method since this variable isn't always used when this is hit.

Also, can we just combine this with `ifNotNullTryParse`? We're already checking `TempSourceStringNotNullExpr` to skip TryParse if null, so we could do this in the else branch in the non-optional case in the same place where we'd set a default value.

Also, does this ever cause a request to fail when a non-optional parameter has a non-null default value? It looks like `int id = 42` would fail for requests without an "id".

Reply:
> Also, does this ever cause a request to fail when a non-optional parameter has a non-null default value?

If a parameter has a default value, it's considered optional by default. `string defaultValueParseableQueryParam(int age = 12) => $"Age: {age}";` validates the scenarios defined here.

> Also, can we just combine this with ifNotNullTryParse? We're already checking TempSourceStringNotNullExpr to skip TryParse if null, so we could do this in the else branch in the non-optional case in the same place where we'd set a default value.

Yeah, it felt a little cleaner to define this in a separate block though. Makes it easier to reason about what check is being made here.

The generated code does look a little weird with the two mutually exclusive if-statements but I think it makes the code to read here a little bit easier.

```
if (string == null)
{
  throw;
}

if (string != null)
{
  // tryParseLogic goes here
}
```

Reply:
As a reviewer, this would have been easier for me to follow if we null checked tempSourceString once per parameter.